### PR TITLE
Improve FermionicOp and SpinOp

### PR DESCRIPTION
**Pull request by ikkoham**
### Summary

1. Fix the but about an empty string (`FermionicOp("")` is not allowed now.)
1. Improve test with assertFemionEqual 
2. ~Deprecate initialization with tuple (`FermionicOp(("I+", coeff))`). Users should use `coeff * FermionicOp("I+")` or `FermionicOp([("I+", coeff)])` instead.~
3. Support sparse label for FermionicOp
4. Sparse label requires register_length.
4. Make internal `FermionicOp._coeff_list` np.ndarray
6. ~Implement `FermionicOp.to_normal_order_list()`~ Move to #88.

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
In regards of 3 above, often we have APIs that take a List[x] (where x in this case is a Tuple[str, complex]) and to simplify things for a user, where they have only one element, so they do not have to make a List to pass it in, we support x directly as a type. In this case you had that and have removed it. Instead requiring the user to make it into a List or to break the Tuple apart to use a different style to initialize it where presumably by passing a str the coeff is implied to be 1 and then multiply it afterwards. Having it support the Tuple directly seems consistent with the usability we do elsewhere. What is the reason to remove it?

**Comment by ikkoham** [comment]
Thanks. For 3, I thought it is just adding square brackets to make a list. The reason for deprecating it is that it would be confusing if there were various initialization methods, and I wanted to save tuple for initializing directly from internal data in the future. Of course, this (latter) isn't necessary because we can detect the type from the element of tuple. If you want to use `Tuple[label, coeff]`, I will revert it.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:101]: I agreee with @woodsp-ibm and would like to keep the `Tuple` option available here :+1:

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:85]: Is there a particular reason for this limitation?

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:101]: OK

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:85]: Yes. Because "+_1 -_0" is not equal to "-_0 +_1".

Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:84]: Exactly. So why can't we allow the latter?

Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:84]: Ah, I got what you said.

It is possible to implement it. But, I think it's non-trivial behavior to initialize an operator that is different from the input label.
Above case, `FermionicOp([('+-', -1)])` is generated.
If user wants to do that,  write explicitly
```
FermionicOp("I_1 -_0") @ FermionicOp("+_1")
```
for now. ("I_1" is not so cool, I want to solve this in the future...)

Comment (pbark) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:84]: Although for the purpose of generality it would be good to support any ordering, I think that respecting the ordering here is not bad in view of the commutators algebra that need to be developed further (in another PR). I believe that keeping the indices in descending order will make the reading of the operator easier. Solving the "I_1" would be nice, as @ikkoham pointed.

**Comment by pbark** [comment]
This LGTM. This should also be reviewed by @woodsp-ibm and after the review we can merge it.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:84]: What exactly needs to happen here to resolve `I_1`? Is that just required in order to make sure that the `register_length` is correct?

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:21]: Why did you change the import here?

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:397]: You are reusing the `list_label` variable here. Please change that.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:411]: I am not sure how I had missed this during the review of the `SpinOp` PR but I assume that this is what you refer to with the `I_1` problem above.
I think it ought to be necessary that the user can simply specify/overwrite the `register_length` himself when using sparse labels. Otherwise it is rather unintuitive that you need to add `I_{register_length - 1}` in order to encode the register length into a sparse label.
If we do not want to add a constructor argument then we need to document this more properly.

File: test/operators/test_fermionic_op.py
Comment (mrossinek) [inline test/operators/test_fermionic_op.py:33]: I think it may be useful to implement `__eq__` in the `FermionicOp` class instead of this.

**Review by mrossinek — CHANGES_REQUESTED** [review]

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:21]: absolute import is better than relative import

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:397]: oh my god. thanks.

File: test/operators/test_fermionic_op.py
Comment (ikkoham) [reply test/operators/test_fermionic_op.py:33]: This is not true equality. Users will be confused.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (woodsp-ibm) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:397]: I am surprised that unit tests did not catch this

**Comment by woodsp-ibm** [comment]
FYI: In commenting on unit test above I will note that in #60 I added a unit test for qubit_converter, which is in operators/second_quant.. and in doing so I created second_quant... folder in that PR and moved the 2 existing unit tests there (which includes the one modified here). That PR is not yet merged so we will have to reconcile things between these.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:84]: We can choose:

1. Explicit: `register_length`
2. Implicit: automatically expansion to larger register_length operator

I prefer the explicit one from the perspective of the Zen of Python.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:397]: But this works correctly in Python and we can not make a test to catch this.

This is same with
```python
num_list = [0, 1, 1, 2, 3]
for num in num_list:
    num_list = [1000]
    print(num)
```
outputs 
```
0
1
1
2
3
```. 
Not 1000.

Of course, I will fix it because readability is bad.

File: test/operators/second_quantization/test_fermionic_op.py
Comment (mrossinek) [inline test/operators/second_quantization/test_fermionic_op.py:81]: What does `targ` stand for?

File: test/operators/second_quantization/test_spin_op.py
Comment (mrossinek) [inline test/operators/second_quantization/test_spin_op.py:68]: Please add some `test_init_from_dense`

File: test/operators/second_quantization/test_spin_op.py
Comment (mrossinek) [inline test/operators/second_quantization/test_spin_op.py:73]: I also don't see the `Tuple`-init case being covered. In fact, I think it may be broken. Please check.

**Review by mrossinek — CHANGES_REQUESTED** [review]

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:217]: These are different. `len` is the number of terms.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:251]: Yes. Thanks. This change is introduced when tuple is deprecated. I forgot to revert it.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:232]: Thnaks. And tolist() is necessary here.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:337]: if tolist() is called, the type of coeffs is complex, but if not called, the type is np.complex128.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:384]: No. We can't. For example, 
```python
ls = [["I"] * 2] * 3
ls[2][1] = 2
print(ls)
```
[['I', 'X'], ['I', 'X'], ['I', 'X']]
```python
ls = [["I"] * 2 for _ in range(3)]
ls[2][1] = 2
print(ls)
```
[['I', 'I'], ['I', 'X'], ['I', 'I']]

File: test/operators/second_quantization/test_fermionic_op.py
Comment (ikkoham) [reply test/operators/second_quantization/test_fermionic_op.py:81]: target. In terra, this is used (https://github.com/Qiskit/qiskit-terra/search?q=targ)

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:232]: Why is that necessary here?

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:384]: Ah right.. We don't get new instances but the same instance `n` times.. Thanks :+1:

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/fermionic_op.py:337]: Alright, this also explains my question above :+1:

File: qiskit_nature/operators/second_quantization/spin_op.py
Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/spin_op.py:224]: `register_length` is valid only for the sparse label as a required parameter, as explained in the documentation.

File: qiskit_nature/operators/second_quantization/fermionic_op.py
Comment (woodsp-ibm) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:61]: In this case register_length is inferred from the label length. I guess its a mistake to state it explicitly, when they are dense labels, as its not going to work?

File: qiskit_nature/operators/second_quantization/spin_op.py
Comment (woodsp-ibm) [inline qiskit_nature/operators/second_quantization/spin_op.py:467]: Its nice/helpful to have the value(s) being referred to be in the message as well ie index and register_length, though it might be more meaningful to a end user to have the whole label there so they can see it in context better

File: qiskit_nature/operators/second_quantization/spin_op.py
Comment (woodsp-ibm) [inline qiskit_nature/operators/second_quantization/spin_op.py]: Similalrly here (and below) the failing label value in the message would be helpful

File: qiskit_nature/operators/second_quantization/spin_op.py
Comment (mrossinek) [inline qiskit_nature/operators/second_quantization/spin_op.py:224]: This does not work as intended, as far as I can tell. Some examples which fail:
```
SpinOp('+_0 i_0')
SpinOp('+_0')  -> this will not be affected by `_flatten_raising_lowering_ops` but then goes through `_flatten_ladder_ops` (which only knows how to handle dense ops)
```

Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/spin_op.py:224]: > SpinOp('+_0 i_0')

This must be error. What is i_0?

> SpinOp('+_0')

Error (register_length is required)

> SpinOp('+_0', register_length=1)

Correct.
`SpinOp([('X_0', (1+0j)), ('Y_0', 1j)], spin=1/2, register_length=1)`

Comment (mrossinek) [reply qiskit_nature/operators/second_quantization/spin_op.py:224]: Ah my apologies. The `register_length` fixes it.
But the error was kind of unexpected because it complains about the invalid labels for dense labels (In the `+_0` case) *after* converting it into `X_0 Y_0`. Maybe we could do an assertion of the labels before that.

Comment (ikkoham) [reply qiskit_nature/operators/second_quantization/spin_op.py:224]: OK I will add the error message somewhere.

**Comment by ikkoham** [comment]
Please merge this PR if normal_order_list() is ok because this PR blocks other PRs. I will add additional tests and improvement of error messages in follow up PR.

**Comment by mrossinek** [comment]
> Please merge this PR if normal_order_list() is ok because this PR blocks other PRs. I will add additional tests and improvement of error messages in follow up PR.

I understand you want to get this out. But, let us wait for the offline decision with regards to the endianess convention.
Which PRs are blocked by this one? I just closed the p/h transformer and I know that `VibrationalProblem` needs this. Anything else?

**Comment by ikkoham** [comment]
Yes. PH and VibrationalProblem.

**Review by mrossinek — APPROVED** [review]
Let's merge this PR today in order to unlock the further development of features which are blocked by this.

*Warning*: the sparse labels are still using little endian. A follow-up PR will fix this!

@ikkoham please also address the remaining open comments of this PR (mainly improvements to error messages as well as additional unittests) in another follow-up PR.

**Comment by ikkoham** [comment]
> Warning: the sparse labels are still using little endian. A follow-up PR will fix this!

Dense label is also little endian now. I've implemented all operators as little endian from #25 and #35.

**Comment by ikkoham** [comment]
Anyway thanks a million. @mrossinek

**Comment by mrossinek** [comment]
> Dense label is also little endian now. I've implemented all operators as little endian from #25 and #35.

Interesting I had not noticed.. I would have assumed that the mapper unittests would fail in this case :thinking:
But let's still proceed for now. The switch to big endian should follow-up early this week anyways :+1:

**Comment by ikkoham** [comment]
Sorry. I was wrong. #25 does not assume endian (endian does not influence) and #35 is little e.g. https://github.com/Qiskit/qiskit-nature/pull/35/files#diff-e1fad0f243ef03387da83f3f9c9b006212b413a01addd08569d3d78048469dd1R131-R132.

**Review by pbark — APPROVED** [review]

**Comment by pbark** [comment]
I think now this LGTM. We need to add the normal ordering in a different PR for future reference and resolve the big endian discussion we had offline (different PR)

**Pull request merged**
